Add support for ploomber engine (#1101)
ploomber/ploomber: doc/cookbook/engine.rst

Engines
=======

Ploomber currently supports 2 engines for running notebooks:

- Ploomber Engine
- Papermill Engine

> `Ploomber Engine <https://engine.ploomber.io/en/latest/quick-start.html>`_

Reviewer: please add sample YAMLs of how choosing one or another would look like.

Also show how to pass parameters to them.

> Engines

Reviewer: ```suggestion
Notebook executors
```

(let's also rename the file to nb_executors.rst)

Reviewer: sorry for the naming changes. I realized that "engines" is a confusing name. it's better to call this "executors" to be consistent with the "NotebookExecutor" name.

so let's rename from `engine` -> `executor`, and `nb_executor_params` -> `executor_params`

=======

> Ploomber currently supports 2 engines for running notebooks:

Reviewer: ```suggestion
Ploomber currently supports two notebook executors:
```

- Ploomber Engine

> - Papermill Engine

Reviewer: ```suggestion
- Papermill (default)
```

let's add a link to papermill's here

also put papermill as the first element since it's the default

> `Ploomber Engine <https://engine.ploomber.io/en/latest/quick-start.html>`_

Reviewer: let's move this link to line 6 where you're listing the engines

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Ploomber Engine is an in-house engine developed on top of Papermill, 
> with better support for error debugging and pipeline deployment. 
> Developers can use Ploomber Engine with Ploomber by setting the ``ploomber_engine`` argument to ``True`` 
> in ``NotebookRunner``. Additionally, developers can pass the following `arguments <https://engine.ploomber.io/en/latest/api/api.html#execute-notebook>`_ in ``nb_executor_params`` while using ``NotebookRunner`` with ``ploomber-engine``.

Reviewer: ploomber-engine isn't built on top of papermill

```suggestion
Ploomber Engine is a notebook executor developer by the Ploomber team with better support for debugging and deployment. 
You can use Ploomber Engine with Ploomber by setting the ``engine`` argument to ``ploomber-engine`` 
in ``NotebookRunner``. Additionally, you can pass the following `arguments <https://engine.ploomber.io/en/latest/api/api.html#execute-notebook>`_ in ``nb_executor_params`` while using ``NotebookRunner`` with ``ploomber-engine``.
```

> 

Reviewer: let's add the yaml example here, something like

```yaml
tasks:
  - source: fit.py
    executor: ploomber-engine
    product:
        nb: output/nb.html
        model: output/model.pickle
    executor_params:
      log_output: True
```

> `Papermill <https://papermill.readthedocs.io/en/latest/>`_

Reviewer: I'm confused about this part:

```
By default, if the ``executor`` argument is not specified in 
``NotebookRunner``, and ``ploomber_engine`` is set to ``False``, 
Ploomber will use Papermill to execute notebooks.
```

does this mean we have *two* options for selecting the executor/engine? we should only have one

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Papermill is a Python library that provides functionality for 
> parameterizing, executing, and analyzing Jupyter notebooks. 
> With Papermill, developers can programmatically interact with notebooks, allowing 
> for automation and customization of notebook workflows. Additionally, 
> Papermill supports custom engines, which can be built on top of its APIs, 
> and has been used for ``debugging`` in Ploomber. Please refer to the short `guide <https://docs.ploomber.io/en/latest/cookbook/debugging.html>`_ for more details on debugging.
> 
> By default, if the ``engine_name`` argument is not specified in 
> ``NotebookRunner``, and ``ploomber_engine`` is set to ``False``, 
> Ploomber will use Papermill to execute notebooks.
> 
> Additionally, developers can pass the following `Papermill arguments <https://papermill.readthedocs.io/en/latest/reference/papermill-workflow.html?highlight=execute_notebook#module-papermill.execute>`_ in ``nb_executor_params`` while using ``NotebookRunner``.
> 
> **Note**: When migrating from one engine to another, make sure to check the parameters 
> passed in ``nb_executor_params``, as different engines might have different arguments and functionalities.

Reviewer: shortening papermill's description, no need to add that many details

```suggestion
Papermill allows you to parameterizing, executing, and analyzing Jupyter notebooks. 

By default, if the ``executor`` argument is not specified in 
``NotebookRunner``, Ploomber will use Papermill to execute notebooks.

Additionally, you can pass the following `Papermill arguments <https://papermill.readthedocs.io/en/latest/reference/papermill-workflow.html?highlight=execute_notebook#module-papermill.execute>`_ in ``executor_params`` while using ``NotebookRunner``.

**Note**: When migrating from one engine to another, make sure to check the parameters 
passed in ``executor_params``, as different engines might have different arguments and functionalities.
```

> Sample pipeline

Reviewer: let's delete this example to keep it short. the sample pipeline should be next to each of the engines

> Sample pipeline

Reviewer: let's delete this example to keep it short. the sample pipeline should be next to each of the runners
